### PR TITLE
Control plane node join

### LINKED ISSUE #413: Add machine actuator support for control plane join
/kind feature

**Describe the solution you'd like**

The machine actuator should support joining additional control plane instances.

Depends on #412

### PR DESCRIPTION
**Pull request by ashish-amarnath**
**What this PR does / why we need it**:
Implements support for control plane node join

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #413

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
cluster-api-provider-aws now supports creating clusters with HA control plane using clusterctl phases.
Steps to create clusters with HA control plane are documented <Link to readme generated as a fix for #414 >
```

### CODE REVIEW

**Comment by ashish-amarnath** [comment]
@chuckha  Thanks for the pointers on #413 
I took a stab at this to get the scaffolding and initial ideas in place.
Publishing this PR to get some feedback on the approach.

/assign @chuckha 
cc @detiber

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 1]
```diff
@@ -78,29 +79,24 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
-		coreClient, err := corev1.NewForConfig(clientConfig)
-		if err != nil {
-			return errors.Errorf("failed to initialize new corev1 client: %+v", err)
-		}
-
-		bootstrapToken, err = tokens.NewBootstrap(coreClient, 10*time.Minute)
-		if err != nil {
-			return errors.Errorf("failed to create new bootstrap token: %+v", err)
+	case "controlplane":
+		_, err := a.GetKubeConfig(cluster, nil)
```
Comment (ashish-amarnath) [inline pkg/cloud/aws/actuators/machine/actuator.go]: This is not a correct check for a control plane node join.

File: pkg/apis/awsprovider/v1alpha1/awsclusterproviderstatus_types.go [hunk 2]
```diff
@@ -30,8 +30,9 @@ type AWSClusterProviderStatus struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Network Network  `json:"network,omitempty"`
-	Bastion Instance `json:"bastion,omitempty"`
+	Network            Network                   `json:"network,omitempty"`
+	Bastion            Instance                  `json:"bastion,omitempty"`
+	ControlPlaneStatus ClusterControlPlaneStatus `json:"clusterControlPlaneStatus,omitempty"`
```
Comment (vincepri) [inline pkg/apis/awsprovider/v1alpha1/awsclusterproviderstatus_types.go]: Is the `Status` suffix necessary here?

File: pkg/apis/awsprovider/v1alpha1/types.go [hunk 3]
```diff
@@ -102,6 +102,12 @@ type Network struct {
 	APIServerELB ClassicELB `json:"apiServerElb,omitempty"`
 }
 
+// ClusterControlPlaneStatus encapsulates control-plane status for the cluster
+type ClusterControlPlaneStatus struct {
+	// ControlPlaneNodeCount is the count of the number of control-plane nodes in the cluster.
+	ControlPlaneNodeCount int `json:"controlPlaneNodeCount"`
```
Comment (vincepri) [inline pkg/apis/awsprovider/v1alpha1/types.go]: Should this just be `NodeCount` given that the struct already sets the context?

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 4]
```diff
@@ -131,6 +123,34 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 	return nil
 }
 
+func (a *Actuator) getWorkerNodeToken(cluster *clusterv1.Cluster, controlPlaneURL string) (string, error) {
+	var bootstrapToken string
```
Comment (vincepri) [inline pkg/cloud/aws/actuators/machine/actuator.go]: Can this value be declared at the bottom? Doesn't seem necessary to be declared here, but I might be wrong.

File: pkg/cloud/aws/actuators/machine_scope.go [hunk 5]
```diff
@@ -128,4 +129,7 @@ func (m *MachineScope) Close() {
 	if err != nil {
 		klog.Errorf("[machinescope] failed to store provider status for machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
 	}
+
+	// control plane node count is a status of the cluster. update that incase a new control plane node was created.
+	m.Scope.Close()
```
Comment (vincepri) [inline pkg/cloud/aws/actuators/machine_scope.go]: This was removed because it causes races between the machine and cluster actuator.

File: pkg/cloud/aws/actuators/scope.go [hunk 6]
```diff
@@ -88,22 +88,27 @@ type Scope struct {
 	ClusterStatus *v1alpha1.AWSClusterProviderStatus
 }
 
+// ClusterControlPlaneStatus returns the cluster controlplane status object.
+func (s *Scope) ClusterControlPlaneStatus() *v1alpha1.ClusterControlPlaneStatus {
```
Comment (vincepri) [inline pkg/cloud/aws/actuators/scope.go]: Could this be simplified just to `ControlPlane()`?

File: pkg/cloud/aws/services/ec2/instances.go [hunk 7]
```diff
@@ -138,41 +139,55 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
...
-			CAKey:             string(s.scope.ClusterConfig.CAPrivateKey),
-			ELBAddress:        s.scope.Network().APIServerELB.DNSName,
-			ClusterName:       s.scope.Name(),
-			PodSubnet:         s.scope.Cluster.Spec.ClusterNetwork.Pods.CIDRBlocks[0],
-			ServiceSubnet:     s.scope.Cluster.Spec.ClusterNetwork.Services.CIDRBlocks[0],
-			ServiceDomain:     s.scope.Cluster.Spec.ClusterNetwork.ServiceDomain,
-			KubernetesVersion: machine.Machine.Spec.Versions.ControlPlane,
-		})
+		var userData string
+		var err error
```
Comment (vincepri) [inline pkg/cloud/aws/services/ec2/instances.go]: Why were these moved here?

File: pkg/cloud/aws/services/ec2/instances.go [hunk 8]
```diff
@@ -138,41 +139,55 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
...
+				KubernetesVersion: machine.Machine.Spec.Versions.ControlPlane,
+			})
 
-		if err != nil {
-			return input, err
+			if err != nil {
+				return input, err
+			}
 		}
+		controlPlaneNodeCount++
```
Comment (vincepri) [inline pkg/cloud/aws/services/ec2/instances.go]: Should we make these a function on the ControlPlane status instead of modifying the value?

File: pkg/cloud/aws/services/ec2/instances.go [hunk 9]
```diff
@@ -204,6 +224,7 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
 	if err != nil {
 		return nil, err
 	}
+	machine.ClusterControlPlaneStatus().ControlPlaneNodeCount = controlPlaneNodeCount
```
Comment (vincepri) [inline pkg/cloud/aws/services/ec2/instances.go]: Same as the other comment, if the increments/decrements are functions we wouldn't need to do this assignment.

File: pkg/cloud/aws/actuators/machine/actuator.go (hunk 4 above)
Comment (ashish-amarnath) [reply pkg/cloud/aws/actuators/machine/actuator.go]: This was a copy pasta :)

File: pkg/apis/awsprovider/v1alpha1/awsclusterproviderstatus_types.go (hunk 2 above)
Comment (ashish-amarnath) [reply pkg/apis/awsprovider/v1alpha1/awsclusterproviderstatus_types.go]: Not taking this approach anymore.

File: pkg/apis/awsprovider/v1alpha1/types.go (hunk 3 above)
Comment (ashish-amarnath) [reply pkg/apis/awsprovider/v1alpha1/types.go]: Dropping this approach

File: pkg/cloud/aws/actuators/machine_scope.go (hunk 5 above)
Comment (ashish-amarnath) [reply pkg/cloud/aws/actuators/machine_scope.go]: Dropping this approach

File: pkg/cloud/aws/services/ec2/instances.go (hunk 7 above)
Comment (ashish-amarnath) [reply pkg/cloud/aws/services/ec2/instances.go]: To share `userData` between controlplane create and controlplane join

File: pkg/cloud/aws/services/ec2/instances.go (hunk 8 above)
Comment (ashish-amarnath) [reply pkg/cloud/aws/services/ec2/instances.go]: Dropping this approach

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 10]
```diff
@@ -131,6 +123,34 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
+	kubeConfig, err := a.GetKubeConfig(cluster, nil)
+	if err != nil {
+		return bootstrapToken, errors.Errorf("failed to retrieve kubeconfig during machine creation: %+v", err)
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply pkg/cloud/aws/actuators/machine/actuator.go]: I'll do a sweep for this once this PR is in the closer to the end.

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 11]
```diff
@@ -131,6 +123,34 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
+	}
+
+	coreClient, err := corev1.NewForConfig(clientConfig)
+	if err != nil {
+		return bootstrapToken, errors.Errorf("failed to initialize new corev1 client: %+v", err)
+	}
+
+	bootstrapToken, err = tokens.NewBootstrap(coreClient, 10*time.Minute)
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply pkg/cloud/aws/actuators/machine/actuator.go]: changed

**Comment by ashish-amarnath** [comment]
Hitting this failure with `kubeadm join`
```
[   13.081046] cloud-init[1303]: kubeadm join --config /tmp/kubeadm-controlplane-join-config.yaml --experimental-control-plane
[   13.233587] cloud-init[1303]: panic: runtime error: invalid memory address or nil pointer dereference
```
ToDo:
* make `--experimental-control-plane` part of the join configuration
* get this to run successfully

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 12]
```diff
@@ -59,6 +61,51 @@ func NewActuator(params ActuatorParams) *Actuator {
...
+	var cpm []*clusterv1.Machine
+	for _, m := range machineList.Items {
+		if m.Spec.Versions.ControlPlane != "" {
+			cpm = append(cpm, m.DeepCopy())
+		}
+	}
+	return cpm
+}
+
+func isMachinesSame(m1 *clusterv1.Machine, m2 *clusterv1.Machine) bool {
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: Rename this to `machinesEqual`.

Add a documentation note that we are defining equality as name and namespace are equivalent and that we are not checking any other fields.

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 13]
```diff
@@ -59,6 +61,51 @@ func NewActuator(params ActuatorParams) *Actuator {
...
+	return m1.Name == m2.Name && m1.Namespace == m2.Namespace
+}
+
+func (a *Actuator) isNodeJoin(controlPlaneMachines []*clusterv1.Machine, newMachine *clusterv1.Machine) (join bool, err error) {
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: don't use named return values here

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 14]
```diff
@@ -59,6 +61,51 @@ func NewActuator(params ActuatorParams) *Actuator {
...
+	err = nil
+	switch newMachine.ObjectMeta.Labels["set"] {
+	case "node":
+		join = true
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: `return true, nil`

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 15]
```diff
@@ -59,6 +61,51 @@ func NewActuator(params ActuatorParams) *Actuator {
...
+	case "controlplane":
+		join = true
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: delete this and directly return the values instead

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 16]
```diff
@@ -59,6 +61,51 @@ func NewActuator(params ActuatorParams) *Actuator {
...
+		// join = false, if:
+		//		1. len(controlPlaneMachines) == 1 && controlPlaneMachines[0] == newMachine
+		if len(controlPlaneMachines) == 1 && isMachinesSame(controlPlaneMachines[0], newMachine) {
+			join = false
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: directly return false and nil here. after this `if` block, we can directly return true and nil instead of setting default values and implicitly returning

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 17]
```diff
@@ -59,6 +61,51 @@ func NewActuator(params ActuatorParams) *Actuator {
...
+		}
+		// 		TODO: ashish-amarnath 2. if none of the controlPlaneMachines exist
+
+		klog.V(2).Infof("Machine %q should join the controlplane: %t", newMachine.Name, join)
+	default:
+		errMsg := fmt.Sprintf("Unknown value %q for label \"set\" on machine %q, skipping machine creation", newMachine.ObjectMeta.Labels["set"], newMachine.Name)
+		klog.Errorf(errMsg)
+		err = errors.Errorf(errMsg)
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: have an explicit return here instead of assigning err

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 18]
```diff
@@ -77,33 +124,31 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
+		return errors.Errorf("failed to retrieve machines for cluster %q: %v", cluster.Name, err)
+	}
+	controlPlaneMachines := a.getControlPlaneMachines(clusterMachines)
+	isNodeJoin, err := a.isNodeJoin(controlPlaneMachines, machine)
+	if err != nil {
+		return errors.Errorf("Failed to determine whther machine %q should join cluster %q: %v", machine.Name, cluster.Name, err)
+	}
 
-		coreClient, err := corev1.NewForConfig(clientConfig)
+	bootstrapToken := ""
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: let's keep this as `var bootstrapToken string`

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 19]
```diff
@@ -77,33 +124,31 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
+	if isNodeJoin {
+		bootstrapToken, err = a.getNodeJoinToken(cluster, controlPlaneURL)
 		if err != nil {
-			return errors.Errorf("failed to initialize new corev1 client: %+v", err)
+			klog.Errorf("failed to retrieve token to create machine %q: %v", machine.Name, err)
```
Comment (chuckha) [inline pkg/cloud/aws/actuators/machine/actuator.go]: I don't think we need klog here, we can instead return the error

File: pkg/cloud/aws/services/ec2/instances.go [hunk 20]
```diff
@@ -138,48 +139,69 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
...
+			userData, err = userdata.JoinControlPlane(&userdata.ContolPlaneJoinInput{
+				CACert:         string(s.scope.ClusterConfig.CACertificate),
+				CAKey:          string(s.scope.ClusterConfig.CAPrivateKey),
+				CACertHash:     caCertHash,
+				BootstrapToken: bootstrapToken,
+				ELBAddress:     s.scope.Network().APIServerELB.DNSName,
+				KubeConfig:     kubeConfig,
+			})
+			if err != nil {
+				return input, err
```
Comment (chuckha) [inline pkg/cloud/aws/services/ec2/instances.go]: make sure this err is wrapped in userdata call

File: pkg/cloud/aws/services/ec2/instances.go [hunk 21]
```diff
@@ -191,6 +213,11 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
 		}
 
 		input.UserData = aws.String(userData)
+
+	default:
+		errMsg := fmt.Sprintf("Unknown node role %q", machine.Role())
+		klog.Error(errMsg)
+		return nil, fmt.Errorf(errMsg)
```
Comment (chuckha) [inline pkg/cloud/aws/services/ec2/instances.go]: we don't need to klog this deep, just return an error and please use the `github.com/pkg/errors` library

File: pkg/cloud/aws/services/userdata/controlplane.go [hunk 22]
```diff
@@ -55,6 +55,49 @@ nodeRegistration:
 EOF
 
 kubeadm init --config /tmp/kubeadm.yaml
+
+tar -cvzf /etc/kubernetes/pki.tar.gz /etc/kubernetes/pki/*
+kubectl -n kube-system --kubeconfig /etc/kubernetes/admin.conf create secret generic kubeadm-certs --from-file=/etc/kubernetes/pki.tar.gz
```
Comment (detiber) [inline pkg/cloud/aws/services/userdata/controlplane.go]: This is an interesting approach for managing the secrets, but it would probably be better if we took a different approach. I can see a couple of different ways of solving this without adding a secret post-init:

- pre-generate the etcd ca keypair, front-proxy ca keypair, and sa(service account) keypair in the cluster controller and store in the Cluster ProviderSpec
  - inject them into the init config like we are currently doing with the CA cert/key
  - pre-write the keypairs to the filesystem before calling kubeadm join
  - if other certificates are needed prior to calling kubeadm join, pregenerate them using kubeadm phases (kubeadm join should only really require the CAs, it would be a bug otherwise)
- Instead of modifying the cluster actuator, it would be possible to fetch the required certificates using the kubernetes client from within the machine actuator, then inject the certificates using the init script on the host prior to running kubeadm join.

Comment (chuckha) [reply pkg/cloud/aws/services/userdata/controlplane.go]: i like the simplicity of this approach. can you elaborate on why you think it would be better if we took a different approach?

i'm guessing security issues? secrets aren't actually secret and it would be far too easy to get these secrets and do bad things. But it's not that different if we store the values in the cluster provider spec. That's still plaintext secret storage in etcd.

Comment (randomvariable) [reply pkg/cloud/aws/services/userdata/controlplane.go]: I would expect operators to use envelope encryption for secrets in their bootstrap / manager clusters, so it'll be justifiable to store ca certificates as secrets, less so as values in the provider spec.

We can't pre-generate the non-authoritative certs yet because we don't know the IP addresses of the nodes beforehand, and this is particularly important for etcd, but reasonable to do so for the certificate authorities and service account key-pair.

For a solution that will also support autoscaling groups, the ideal way would be use the [instance identity document](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html) as a host attestation mechanism, where a verifiable CSR is made, that the controller can sign.

Comment (detiber) [reply pkg/cloud/aws/services/userdata/controlplane.go]: So, there are a couple of different reasons why I'm suggesting adding it to the provider spec:
- It automatically provides an avenue for users to override the CA and SA keypairs to use custom values
- While we are embedding the certificates in the spec today, we can later update the spec to allow for either embedding or referencing an external source at a later point

I want to make sure that we are considering UX when we start to look at more secure private key storage, by allowing direct embedding of the cert material, we avoid having to pre-generate any additional secrets as part of the workflow and we the spec is updated to support remote references for secure material (k8s secrets or leveraging something AWS native), we can have the actuator move the cert material to the proper place as part of the workflow, or even have an admission webhook that does the job to avoid ever storing the cert material in etcd).

Comment (randomvariable) [reply pkg/cloud/aws/services/userdata/controlplane.go]: Happy with this approach. So we need to change the spec a bit now, at least, to hold all the key pairs we need for the HA control plane?

Comment (ashish-amarnath) [reply pkg/cloud/aws/services/userdata/controlplane.go]: I agree with the UX and the extendability aspect. However, IMO, we have a dependency on `kubeadm` here for generating the certs on when the first controlplane node runs the init. From reading the kubeadm docs, there seems to be a way to skip cert generation and run the other "phases". By doing this we will be deviating from the standard `kubeadm init` flow. Which, IMO, is OK as long as we call that out for UX.

Also, kubeadm will start handling the cert distribution for controlplane join in v.1.14. More info https://github.com/kubernetes/kubeadm/issues/1200
With this in mind and we aren't GA  yet, I think the cert distribution fix I have should be a stop-gap till we go to 1.14. WDYT?

Comment (detiber) [reply pkg/cloud/aws/services/userdata/controlplane.go]: > I agree with the UX and the extendability aspect. However, IMO, we have a dependency on kubeadm here for generating the certs on when the first controlplane node runs the init. From reading the kubeadm docs, there seems to be a way to skip cert generation and run the other "phases". By doing this we will be deviating from the standard kubeadm init flow. Which, IMO, is OK as long as we call that out for UX.

@ashish-amarnath If the keypair files already exist on disk in the default locations that kubeadm would use for creating them and are not close to expiration they will be picked up and used by kubeadm without having to run phases.

> Also, kubeadm will start handling the cert distribution for controlplane join in v.1.14. More info kubernetes/kubeadm#1200
With this in mind and we aren't GA yet, I think the cert distribution fix I have should be a stop-gap till we go to 1.14. WDYT?

Even if kubeadm handles the certificate copying for v1.14+ we would still benefit from having the ability for users to specify custom keypairs here.

Comment (ashish-amarnath) [reply pkg/cloud/aws/services/userdata/controlplane.go]: I've decoupled the certs into their individual secrets so that they can be each updated independent of the other secrets.

Comment (chuckha) [reply pkg/cloud/aws/services/userdata/controlplane.go]: +1 thanks for the reasoning @detiber. totally agree that changing the spec is the right way to go.

Comment (ashish-amarnath) [reply pkg/cloud/aws/services/userdata/controlplane.go]: So, I propose that we keep the spec change out of this PR and get the controlplane machine join scenario worked out. I'll follow this PR with the spec change.

File: pkg/cloud/aws/services/ec2/instances.go (hunk 9 above)
Comment (ashish-amarnath) [reply pkg/cloud/aws/services/ec2/instances.go]: Outdated

File: pkg/cloud/aws/actuators/machine/actuator.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply pkg/cloud/aws/actuators/machine/actuator.go]: I did a sweep to wrap errors.

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 23]
```diff
@@ -131,6 +123,34 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
+	if err != nil {
+		return bootstrapToken, errors.Errorf("failed to retrieve kubeconfig during machine creation: %+v", err)
+	}
+
+	clientConfig, err := clientcmd.BuildConfigFromKubeconfigGetter(controlPlaneURL, func() (*clientcmdapi.Config, error) {
+		return clientcmd.Load([]byte(kubeConfig))
+	})
+
+	if err != nil {
+		return bootstrapToken, errors.Errorf("failed to retrieve kubeconfig during machine creation: %+v", err)
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply pkg/cloud/aws/actuators/machine/actuator.go]: Thanks for explaining! +1
Fixed now.

File: pkg/cloud/aws/services/userdata/controlplane.go (hunk 22 above)
Comment (ashish-amarnath) [reply pkg/cloud/aws/services/userdata/controlplane.go]: Created https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/492 to address this.

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 24]
```diff
@@ -77,33 +130,30 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
-			return errors.Errorf("failed to retrieve kubeconfig during machine creation: %+v", err)
-		}
+	clusterMachines, err := scope.MachineClient.List(v1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve machines in cluster %q", cluster.Name)
+	}
+	controlPlaneMachines := a.getControlPlaneMachines(clusterMachines)
+	isNodeJoin, err := a.isNodeJoin(controlPlaneMachines, machine, cluster)
+	if err != nil {
+		return errors.Wrapf(err, "failed to determine whther machine %q should join cluster %q", machine.Name, cluster.Name)
```
Comment (daniel-garcia) [inline pkg/cloud/aws/actuators/machine/actuator.go]: whether vs whther

**Comment by chuckha** [comment]
let's type better

/approve
